feat: add python3.10 support and make code python3.7+

vinitkumar opened the pull request:
Add python3.10 to the test suite and also ensure the code is python3.7+
in terms on syntax

Authored-by: Vinit Kumar <vinit.kumar@kidskonnect.nl>
Signed-off-by: Vinit Kumar <vinit.kumar@kidskonnect.nl>

## Proposed Changes

Add python3.10 as it has been stable for almost an year now.
Make syntax modern with pyupgrade supporting python3.7+

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask on the
[`pika-python`](https://groups.google.com/forum/#!forum/pika-python)
mailing list. We're here to help! This is simply a reminder of what we
are going to look for before merging your code._

- [ ] I have read the `CONTRIBUTING.md` document
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further Comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc.

lukebakken commented:
@vinitkumar thanks for your contribution. How did you determine what code needed to be updated? Did you run an automated tool?

lukebakken commented:
@vinitkumar thanks for your contribution. How did you determine what code needed to be updated? Did you run an automated tool?

gaby commented:
@lukebakken Probably used this https://github.com/asottile/pyupgrade

gaby commented:
The only problem I see with this PR, is that projects using anything older than py3.7, won't be able to get any more updates. Besides that 👍🏻

gaby commented:
Yup, makes sense! 👍🏻

lukebakken commented:
Thanks for the discussion. Yes, this code will have to wait for Pika 2.0.

vinitkumar commented:
@lukebakken Alrighty, thanks for taking the time.

michaelklishin commented:
@vinitkumar can you please resolve the conflicts this branch now has with `main`? Thank you.

vinitkumar commented:
@michaelklishin Yes, will do that now.

vinitkumar commented:
@michaelklishin This should be ready to review now and you might need to approve the CI run.

I also updated other directories, such as tests, examples and utils that I missed last time.

So right now, if this PR gets merged, all of the code is Python3.7 and above supported with latest syntax in them.

lukebakken commented:
> So right now, if this PR gets merged, all of the code is Python3.7 and above supported with latest syntax in them

Great, thank you. I'll see if 1.3.0 even works with any Pythons before 3.7.

I'm away from work until July 12th and will return to this then.

gopackgo90 commented:
I added a separate PR to add `requires-python` to help users so they don't accidentally install incompatible versions of `pika` in their environments and don't have to pin the versions themselves, https://github.com/pika/pika/pull/1376. This PR could change that to `requires-python = ">=3.7"`, which would help users not install pika 2.0 on python 3.6 and below.

There are also [places in the README](https://github.com/pika/pika/blame/1.3.0/README.rst#L12) that mention pika being installable on python 3.4+ as well as the [trove classifiers](https://github.com/pika/pika/blob/1.3.0/pyproject.toml#L21-L23) that should be addressed.

vinitkumar commented:
> I added a separate PR to add `requires-python` to help users so they don't accidentally install incompatible versions of `pika` in their environments and don't have to pin the versions themselves, #1376. This PR could change that to `requires-python = ">=3.7"`, which would help users not install pika 2.0 on python 3.6 and below.
> 
> There are also [places in the README](https://github.com/pika/pika/blame/1.3.0/README.rst#L12) that mention pika being installable on python 3.4+ as well as the [trove classifiers](https://github.com/pika/pika/blob/1.3.0/pyproject.toml#L21-L23) that should be addressed.

Thanks for the tip. I ensure that the library is installable only on Python3.7+ and also updated the readme.

michaelklishin commented:
FTR, this has not been forgotten but it's high PTO season and I don't want to be the only person who decides that this is ready to go :)

vinitkumar commented:
@michaelklishin Thanks for taking the time to reply. Just one request, can you please approve the workflow run so that I can see whether any CI fails or not?

vinitkumar commented:
@lukebakken @gaby @michaelklishin I just fixed some conflicts with the main and pushed the latest commits for it. 
Please approve the CI run and see if this can be merged into the codebase? 

Thanks for taking the time.

vinitkumar commented:
Are there any blockers for this getting merged?

lukebakken commented:
I'm on vacation and will return to this next week.

Does anyone urgently need these changes?

On Wed, Oct 12, 2022, 06:00 Vinit Kumar ***@***.***> wrote:

> Are there any blockers for this getting merged?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pika/pika/pull/1351#issuecomment-1276138412>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD5W3WYH6W7W3A4537ACZDWC2Y7FANCNFSM5TS5P2DA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

vinitkumar commented:
Thank you for your reply. Well, it has been lying for months now and hence
asked if there is an intention to merge this.

On Wed, 12 Oct 2022 at 18:48, Luke Bakken ***@***.***> wrote:

> I'm on vacation and will return to this next week.
>
> Does anyone urgently need these changes?
>
> On Wed, Oct 12, 2022, 06:00 Vinit Kumar ***@***.***> wrote:
>
> > Are there any blockers for this getting merged?
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pika/pika/pull/1351#issuecomment-1276138412>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAD5W3WYH6W7W3A4537ACZDWC2Y7FANCNFSM5TS5P2DA
> >
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pika/pika/pull/1351#issuecomment-1276161437>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEDITSMKX5LBJAK3RFKL2DWC23A5ANCNFSM5TS5P2DA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Thanks and Regards,
Vinit Kumar

lukebakken commented:
> Well, it has been lying for months now and hence asked if there is an intention to merge this.

Of course there is. I said a [while back](https://github.com/pika/pika/pull/1351#issuecomment-1111046306) that these changes must wait for Pika 2.0. What I'll do is make sure that 1.3.1 is done, release that, then I can merge this.

Thanks for your contribution. If you didn't know, the RabbitMQ team maintains this library. We're a small team so things take time.

vinitkumar commented:
> > Well, it has been lying for months now and hence asked if there is an intention to merge this.
> 
> Of course there is. I said a [while back](https://github.com/pika/pika/pull/1351#issuecomment-1111046306) that these changes must wait for Pika 2.0. What I'll do is make sure that 1.3.1 is done, release that, then I can merge this.
> 
> Thanks for your contribution. If you didn't know, the RabbitMQ team maintains this library. We're a small team so things take time.
> 
> This library is used in production by many RabbitMQ users and is very stable, so we make changes _slowly_.

Thank you, and I understand now.

lukebakken commented:
@vinitkumar thanks!
